### DISCUSSION TITLE
Заброшенная шахта

### DISCUSSION THREAD

**Discussion by Kor-Angar** [dosinabox/g2nr_unofficial_update#513] [General]
,,,,

**Comment by Kor-Angar** [comment]
при отключение мода герой продолжает переходить в несуществующий мир. 

в следующей версии может быть было бы лучше сделать переход на скриптах.

**Reply by dosinabox** [reply]
> при отключение мода герой продолжает переходить в несуществующий мир.

Измененные триггеры записываются в сохраненку? Забавно.

> в следующей версии может быть было бы лучше сделать переход на скриптах.

Как ты это представляешь? В неофициальном обновлении в NW и OW у меня стоят триггеры перед зонами перехода, они ссылаются на скрипты и мир можно менять там, да. Но если юзер установит Wasteland или текстурно-модельный пак с измененными мирами, то очередность перехода NW <> AM <> OW будет нарушена. Поэтому и появился плагин - чтобы правильный переход не был зависим от миров или скриптов и работал бы даже в модах и на любой языковой версии оригинала.

**Reply by dosinabox** [reply]
> В неофициальном обновлении в NW и OW у меня стоят триггеры перед зонами перехода, они ссылаются на скрипты и мир можно менять там

Вот тут:
```cpp
func void B_PassedAbandonedMine()
{
	PlayerPassedAbandonedMine = TRUE;
	if(CurrentLevel == NEWWORLD_ZEN)
	{
		Wld_ChangeLevel("ABANDONEDMINE.ZEN","START_NW");
	}
	else if(CurrentLevel == OLDWORLD_ZEN)
	{
		Wld_ChangeLevel("ABANDONEDMINE.ZEN","START_OW");
	};
};
```

**Reply by Kor-Angar** [reply]
> Как ты это представляешь?

в `Game_Loop` проверяется не подошёл ли герой в зону перехода. если подошёл меняется мир. там же есть какие то точки по которым можно сделать проверку.

***

надо или плагин допилить что бы он не оставлял следов и мир можно было отключить. или сделать аналог на скриптах. судя по тому что плагин простой и несёт понятный смысл наверное его можно доделать :)

**Comment by Kor-Angar** [comment]
какие то дырки синие. а чёрные дырки сделать нельзя? :-)

**Reply by Kor-Angar** [reply]
![20230223223512_1](https://user-images.githubusercontent.com/49514764/222024600-8fd52a6c-5e35-424a-a6ff-d378730add48.jpg)
![20230223223423_1](https://user-images.githubusercontent.com/49514764/222024626-8e3ca9e8-c0a8-461e-b697-c0adc36b586a.jpg)
![20230223223302_1](https://user-images.githubusercontent.com/49514764/222024645-8c528d77-5c08-477c-bf18-93e6b2ecf313.jpg)

**Reply by Kor-Angar** [reply]
![20230223230729_1](https://user-images.githubusercontent.com/49514764/222024706-5bbdc175-8098-47aa-98f7-69762fb751f6.jpg)

**Reply by Kor-Angar** [reply]
![20230223223534_1](https://user-images.githubusercontent.com/49514764/222024819-9115e5af-7d84-4e8b-8088-430aa39532ec.jpg)

**Reply by dosinabox** [reply]
исправлено позавчера

**Reply by dosinabox** [reply]
там пустые текстуры были

**Reply by Kor-Angar** [reply]
зер гуд

**Comment by Kor-Angar** [comment]
какие то скриншоты для стима

**Reply by Kor-Angar** [reply]
![20230223224701_1](https://user-images.githubusercontent.com/49514764/222024773-73cc8a38-e6e4-42f1-834f-825c921694a8.jpg)

**Reply by dosinabox** [reply]
для стима надо делать без DX11 и со стандартными шрифтамм

**Comment by Kor-Angar** [comment]
возможные НПС для шахты

**Reply by Kor-Angar** [reply]
![image](https://user-images.githubusercontent.com/49514764/222024934-0761ffe4-f8b5-4b82-a47b-61bc94351468.png)
![image](https://user-images.githubusercontent.com/49514764/222024946-611e9a4b-2e38-4cd8-bd2b-b954a534f688.png)
![image](https://user-images.githubusercontent.com/49514764/222024970-6078016b-e493-48d1-bf29-3339be7e4340.png)

еще кузнец и помощник кузнеца

**Reply by Kor-Angar** [reply]
ну с кого то надо начинать :-) https://www.youtube.com/watch?v=IGXHxWdBZ_s

```D
	func void B_ENTER_ABANDONEDMINE_Kapitel_4()
	{
        if(EnterAM[4] == FALSE)
		{
			EnterAM[4] = TRUE;
            
            // добавить маркеры при смерти НПС в старом мире
            ExchangeRoutineAbandonedMine (STRF_1115_Geppert);  
            ExchangeRoutineAbandonedMine (STRF_1116_Kervo);
        
            ExchangeRoutineAbandonedMine (VLK_4106_Dobar);
            ExchangeRoutineAbandonedMine (VLK_4107_Parlaf);
```

[AM_NPCRoutine.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/10893944/AM_NPCRoutine.zip)

**Comment by dosinabox** [comment]
исходный код плагина:
```cpp
#if ENGINE >= Engine_G2

namespace NAMESPACE
{
    Sub onLoadEnd(ZSUB(GameEvent::LoadEnd), []
        {
            zCWorld* const world = ogame->GetWorld();

            if (!world)
                return;

            zSTRING triggerName;
            zSTRING levelName;
            zSTRING startVob;

            if (world->m_strlevelName.CompareI("NEWWORLD"))
            {
                triggerName = "NW_2_OW_ABANDONEDMINE";
                levelName = "ABANDONEDMINE.ZEN";
                startVob = "START_NW";
            }
            else if (world->m_strlevelName.CompareI("OLDWORLD"))
            {
                triggerName = "OW_2_NW_ABANDONEDMINE";
                levelName = "ABANDONEDMINE.ZEN";
                startVob = "START_OW";
            }
            else
                return;

            if (oCTriggerChangeLevel* trigger = dynamic_cast<oCTriggerChangeLevel*>(world->SearchVobByName(triggerName)))
            {
                trigger->levelName = levelName;
                trigger->startVob = startVob;
            }
        });
}

#endif
```

**Reply by dosinabox** [reply]
Без записи в сохраненку:
```cpp
namespace NAMESPACE
{
	class TemporaryTriggerChange
	{
	private:
		std::vector<Sub<void>> subs;
		ZOwner<oCTriggerChangeLevel> trigger;
		zSTRING levelName;
		zSTRING startVob;

		void OnLoadBegin()
		{
			delete this;
		}

		void OnSaveBegin()
		{
			std::swap(trigger->levelName, levelName);
			std::swap(trigger->startVob, startVob);
		}

		void OnSaveEnd()
		{
			std::swap(trigger->levelName, levelName);
			std::swap(trigger->startVob, startVob);
		}

		void OnExit()
		{
			delete this;
		}

		~TemporaryTriggerChange()
		{
			trigger->levelName = levelName;
			trigger->startVob = startVob;
		}

	public:
		TemporaryTriggerChange(oCTriggerChangeLevel* trigger, const zSTRING& levelName, const zSTRING& startVob) :
			trigger{ trigger },
			levelName{ trigger->levelName },
			startVob{ trigger->startVob }
		{
			trigger->AddRef();
			trigger->levelName = levelName;
			trigger->startVob = startVob;

			ADDSUB(LoadBegin);
			ADDSUB(SaveBegin);
			ADDSUB(SaveEnd);
			ADDSUB(Exit);
		}
	};

	Sub changeTrigger(ZSUB(GameEvent::LoadEnd), []
		{
			oCWorld* const world = ogame->GetGameWorld();

			zSTRING triggerName;
			zSTRING levelName;
			zSTRING startVob;

			if (world->worldName.CompareI("NEWWORLD"))
			{
				triggerName = "NW_2_OW_ABANDONEDMINE";
				levelName = "ABANDONEDMINE.ZEN";
				startVob = "START_NW";
			}
			else if (world->worldName.CompareI("OLDWORLD"))
			{
				triggerName = "OW_2_NW_ABANDONEDMINE";
				levelName = "ABANDONEDMINE.ZEN";
				startVob = "START_OW";
			}
			else
				return;

			if (oCTriggerChangeLevel* trigger = dynamic_cast<oCTriggerChangeLevel*>(world->SearchVobByName(triggerName)))
				new TemporaryTriggerChange{ trigger, levelName, startVob };
		});
}
```

**Reply by Kor-Angar** [reply]
отл.

**Comment by dosinabox** [comment]
Обновлено:
- добавлены недостающие анимации зажигания факелов для оригинала;
- измененные триггеры перехода не будут записываться в сохранения.

**Comment by Kor-Angar** [comment]
положу здесь, что бы была

[ABANDONEDMINE.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/10893442/ABANDONEDMINE.zip)

**Comment by Kor-Angar** [comment]
вставку жуком и лута лучше перенести из `STARTUP_AbandonedMine` в  `B_ENTER_ABANDONEDMINE_Kapitel_1`

**Comment by Kor-Angar** [comment]
не нашёл расписание с зажиганием факела. что то сочинил.

![20230306180918_1](https://user-images.githubusercontent.com/49514764/223150948-67707099-5e39-4d1a-9046-a776055e404a.jpg)

**Reply by Kor-Angar** [reply]
```D
func void ZS_Patrul_FIRE ()
{
	Perception_Set_Normal();
	
	B_ResetAll (self);

	AI_SetWalkmode 	(self, NPC_WALK);		
	
	if (Hlp_StrCmp (Npc_GetNearestWP(self), self.wp) == FALSE)
    {
		AI_GotoWP	(self, self.wp);
	};
};

FUNC int ZS_Patrul_FIRE_Loop ()
{
    if (!C_BodyStateContains(self, BS_MOBINTERACT_INTERRUPT))
    && (Wld_IsMobAvailable(self,"FIREPLACE"))
    {
        if Wld_GetMobState(self,"FIREPLACE") == 0
        {
            Npc_ClearAIQueue (self);

            AI_UseMob (self, "FIREPLACE",  1);
            AI_UseMob (self, "FIREPLACE", -1);
        }
        else    
        {
            // факел Горит, НПС делает что то иное
            AI_StartState(self,ZS_Smoke_Joint,0,self.wp);
        };   
    }; 
    
    return LOOP_CONTINUE;
};

FUNC VOID ZS_Patrul_FIRE_End ()
{

};

func void TA_Patrul_FIRE			(var int start_h, var int start_m, var int stop_h, var int stop_m, VAR string waypoint)	{TA_Min		(self,	start_h,start_m, stop_h, stop_m, ZS_Patrul_FIRE, waypoint);};
```
